libc/stdlib_env.c

```diff
@@ -0,0 +1,449 @@
+ * Author: Krystian Wasik, Adam Debek
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <time.h>
+#include <unity_fixture.h>
+
+#define NUM_OF_VARIABLES (64)
+#define NUM_OF_TESTS     (1000)
+#define MAX_NAME_LEN     (64)
+#define MAX_VALUE_LEN    (32)
+
+/* Defines how often environment should be cleared (approximately) */
+#define CLEARENV_APPROX_EVERY (200)
+
+#define TEST_VAR     ("TEST_VARIABLE")
+#define TEST_VALUE   ("TEST_VALUE")
+#define TEST_VALUE2  ("TEST_VALUE2")
+#define INVALID_VAL1 ("VA=LUE")
+#define INVALID_VAL2 ("=VALUE")
+#define INVALID_VAL3 ("VALUE=")
+#define INVALID_VAR1 ("NAME=")
+#define INVALID_VAR2 ("NA=ME")
+#define INVALID_VAR3 ("=NAME")
+#define NOT_SET_VAR  ("NOT_SET_VAR")
+#define VAR_VAL      ("NAME=VALUE")
+
+
+typedef struct {
+	char name[MAX_NAME_LEN];                /* string containing name of env variable */
+	char value[MAX_VALUE_LEN];              /* string containing value of env variable */
+	char str[MAX_NAME_LEN + MAX_VALUE_LEN]; /* string of the form "name=value" */
+	unsigned int set;                       /* flag indicating if env variable is set */
+} env_var_t;
+
+static env_var_t vars[NUM_OF_VARIABLES];
+
+extern char **environ;
+
+
+TEST_GROUP(stdlib_env);
+
+
+TEST_SETUP(stdlib_env)
+{
+}
+
+
+TEST_TEAR_DOWN(stdlib_env)
+{
+}
+
+
+static int get_environ_len(char **ep)
+{
+	int i = 0;
+	if (ep != NULL)
+		while (*ep != NULL) {
+			++i;
+			++ep;
+		}
+	else {
+		return 0;
+	}
+
+	return i;
+}
+
+
+static char *get_ascii_str(void)
+{
+	char *str = (char *)malloc(128 * sizeof(char));
+	TEST_ASSERT_NOT_NULL(str);
+	int i;
+
+	for (i = 1; i < 128; i++) {
+		if (i == '=') {
+			str[i - 1] = 'a';
+		}
+		else {
+			str[i - 1] = i;
+		}
+	}
+	str[i - 1] = '\0';
+
+	return str;
+}
+
+
+static char *get_long_str(void)
+{
+	char *str = (char *)malloc(1024 * sizeof(char));
+	TEST_ASSERT_NOT_NULL(str);
+	int i;
+
+	for (i = 0; i < 1023; i++) {
+		str[i] = 'a';
+	}
+	str[i - 1] = '\0';
+
+	return str;
+}
+
+
+TEST(stdlib_env, basic_val_name)
+{
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, setenv(TEST_VAR, TEST_VALUE, 0));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	TEST_ASSERT_EQUAL_STRING(TEST_VALUE, getenv(TEST_VAR));
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, putenv(VAR_VAL));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	TEST_ASSERT_EQUAL_STRING("VALUE", getenv("NAME"));
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, unsetenv("NAME"));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	TEST_ASSERT_NULL(getenv("NAME"));
+}
+
+
+TEST(stdlib_env, long_val_name)
+{
+	char *name = get_long_str();
+	char *value = get_long_str();
+	int len = strlen(name) + strlen(value) + 2;
+	char s[len];
+	len = strlen(name);
+
+	strcpy(s, name);
+	s[len] = '=';
+	strcpy(s + (len + 1), value);
+
+	errno = 0;
+	TEST_ASSERT_EQUAL_INT(0, setenv(name, value, 0));
+	TEST_ASSERT_EQUAL_INT(0, errno);
+
+	TEST_ASSERT_EQUAL_STRING(value, getenv(name));
+
```

Comment:
We could call `unsetenv` there to clear the `name` env variable, now we are not sure whether `putenv` worked fine, because it's set to `value` anyway.